flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java

```diff
@@ -0,0 +1,213 @@
+     *     PackagedProgram} instance.
+     * @throws FlinkException If something goes wrong during instantiation.
+     */
+    public static PackagedProgramRetrieverImpl create(
+            @Nullable File userLibDir,
+            @Nullable String jobClassName,
+            String[] programArgs,
+            Configuration configuration)
+            throws FlinkException {
+        return create(userLibDir, null, jobClassName, programArgs, configuration);
+    }
+
+    /**
+     * Creates a {@code PackageProgramRetrieverImpl} with the given parameters.
+     *
+     * @param userLibDir The user library directory that is used for generating the user classpath
+     *     if specified. The system classpath is used if not specified.
+     * @param jarFile The jar archive used expecting to have the job class included.
+     * @param jobClassName The job class that will be used if specified. The classpath is used to
+     *     detect any main class if not specified.
+     * @param programArgs The program arguments.
+     * @param configuration The Flink configuration for the given job.
+     * @return The {@code PackageProgramRetrieverImpl} that can be used to create a {@link
+     *     PackagedProgram} instance.
+     * @throws FlinkException If something goes wrong during instantiation.
+     */
+    public static PackagedProgramRetrieverImpl create(
+            @Nullable File userLibDir,
+            @Nullable File jarFile,
+            @Nullable String jobClassName,
+            String[] programArgs,
+            Configuration configuration)
+            throws FlinkException {
+        List<URL> userClasspath;
+        try {
+            userClasspath = extractUserClasspath(userLibDir);
+        } catch (IOException e) {
+            throw new FlinkException("An error occurred while extracting the user classpath.", e);
+        }
+
+        final EntryClassInformationProvider entryClassInformationProvider =
+                createEntryClassInformationProvider(
+                        userLibDir, userClasspath, jarFile, jobClassName, programArgs);
+        return new PackagedProgramRetrieverImpl(
+                entryClassInformationProvider, programArgs, userClasspath, configuration);
+    }
+
+    @VisibleForTesting
+    static EntryClassInformationProvider createEntryClassInformationProvider(
+            @Nullable File userLibDir,
+            Iterable<URL> userClasspath,
+            @Nullable File jarFile,
+            @Nullable String jobClassName,
+            String[] programArgs)
+            throws FlinkException {
+        if (PackagedProgramUtils.isPython(jobClassName)
+                || PackagedProgramUtils.isPython(programArgs)) {
+            return FromJarEntryClassInformationProvider.createFromPythonJar();
+        }
+
+        if (jarFile != null) {
+            return FromJarEntryClassInformationProvider.createFromCustomJar(jarFile, jobClassName);
+        }
+
+        if (userLibDir != null) {
```

Comment:
this is a bit weird isn't it. Couldn't we make `userClasspath` nullable instead? And move the null check for the userlib to the `create` variant that accepts it? That would also address https://github.com/apache/flink/pull/16286/files/a7a79b1944b91a03c7567ce29c561186b1f13d74..c071a23a319c89310e19a40f6d85217936ab1067#r662129429.

Reply:
I don't know why I didn't see the obvious. 🤦 But yeah, that's way better.